File: scripts/validate_yaml.py
#!/usr/bin/env python3
"""
YAML validation script for AISMM model definitions.
Supports both:
  - v1.x structure (flat domains dictionary, pillars, scoring_config)
  - v2.x structure (aismm wrapper with nested components)
"""
import sys
import os
import yaml

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
YAML_PATH = os.path.join(ROOT, "aismm_definition", "aismm.yaml")


def load_yaml(path):
    """Load YAML file, rejecting files with tab characters."""
    with open(path, "r", encoding="utf-8") as f:
        text = f.read()
    if '\t' in text:
        lines_with_tabs = [i + 1 for i, ln in enumerate(text.splitlines()) if '\t' in ln]
        raise ValueError(
            f"YAML file contains tab characters on line(s): {lines_with_tabs}. "
            "YAML must use spaces for indentation. Replace tabs with spaces and re-run validation."
        )
    return yaml.safe_load(text)


def detect_version(doc):
    """Detect which AISMM structure version is being used."""
    if not isinstance(doc, dict):
        return None
    # v2.x has 'aismm' top-level key with 'components' inside
    if "aismm" in doc:
        return "2.x"
    # v1.x has 'domains', 'pillars', 'scoring_config' at top level
    if "domains" in doc and "pillars" in doc:
        return "1.x"
    return None


def validate_v1(doc):
    """Validate v1.x structure (flat domains, pillars, scoring_config)."""
    errors = []
    
    # Check required top-level keys
    required_keys = ["version", "name", "description", "scoring_config", "pillars", "domains"]
    for key in required_keys:
        if key not in doc:
            errors.append(f"Missing required top-level key: '{key}'")
    
    # Validate scoring_config
    sc = doc.get("scoring_config", {})
    if not isinstance(sc, dict):
        errors.append("'scoring_config' should be a mapping")
    else:
        for key in ["level_scores", "maturity_thresholds"]:
            if key not in sc:
                errors.append(f"scoring_config missing '{key}'")
    
    # Validate pillars
    pillars = doc.get("pillars", {})
    if not isinstance(pillars, dict):
        errors.append("'pillars' should be a mapping")
    else:
        for pid, pdata in pillars.items():
            if not isinstance(pdata, dict):
                errors.append(f"pillars.{pid} should be a mapping")
                continue
            for field in ["id", "name", "description", "weight"]:
                if field not in pdata:
                    errors.append(f"pillars.{pid} missing '{field}'")
    
    # Validate domains
    domains = doc.get("domains", {})
    if not isinstance(domains, dict):
        errors.append("'domains' should be a mapping")
    else:
        for did, ddata in domains.items():
            prefix = f"domains.{did}"
            if not isinstance(ddata, dict):
                errors.append(f"{prefix} should be a mapping")
                continue
            
            # Required domain fields
            for field in ["id", "name", "description", "pillar"]:
                if field not in ddata:
                    errors.append(f"{prefix} missing '{field}'")
            
            # Validate pillar reference
            pillar_ref = ddata.get("pillar")
            if pillar_ref and pillar_ref not in pillars:
                errors.append(f"{prefix}.pillar '{pillar_ref}' not found in pillars definition")
            
            # Validate levels
            levels = ddata.get("levels", {})
            if not isinstance(levels, dict):
                errors.append(f"{prefix}.levels should be a mapping")
            else:
                level_keys = set(levels.keys())
                expected_levels = {"level_1", "level_2", "level_3", "level_4", "level_5"}
                if level_keys != expected_levels:
                    missing = expected_levels - level_keys
                    extra = level_keys - expected_levels
                    if missing:
                        errors.append(f"{prefix}.levels missing: {missing}")
                    if extra:
                        errors.append(f"{prefix}.levels has unexpected keys: {extra}")
            
            # Validate framework_alignment (optional but recommended)
            fa = ddata.get("framework_alignment", [])
            if fa and not isinstance(fa, list):
                errors.append(f"{prefix}.framework_alignment should be a list")
            
            # Validate key_controls (optional but recommended after merge)
            kc = ddata.get("key_controls", [])
            if kc and not isinstance(kc, list):
                errors.append(f"{prefix}.key_controls should be a list")
            
            # Validate questions
            questions = ddata.get("questions", [])
            if questions and not isinstance(questions, list):
                errors.append(f"{prefix}.questions should be a list")
            elif questions:
                for qi, q in enumerate(questions, start=1):
                    qprefix = f"{prefix}.questions[{qi}]"
                    if not isinstance(q, dict):
                        errors.append(f"{qprefix} should be a mapping")
                        continue
                    for field in ["id", "text", "question_type"]:
                        if field not in q:
                            errors.append(f"{qprefix} missing '{field}'")
    
    return errors


def validate_v2(doc):
    """Validate v2.x structure (aismm wrapper with nested components)."""
    errors = []
    
    if "aismm" not in doc:
        errors.append("Missing top-level 'aismm' key")
        return errors
    
    aismm = doc.get("aismm", {})
    if not isinstance(aismm, dict):
        errors.append("'aismm' should be a mapping")
        return errors
    
    comps = aismm.get("components")
    if comps is None:
        errors.append("Missing 'components' list")
    elif not isinstance(comps, list):
        errors.append("'components' should be a list")
    else:
        for ci, comp in enumerate(comps, start=1):
            prefix = f"components[{ci}]"
            if not isinstance(comp, dict):
                errors.append(f"{prefix} is not a mapping")
                continue
            if "id" not in comp:
                errors.append(f"{prefix} missing 'id'")
            if "id_code" not in comp:
                errors.append(f"{prefix} missing 'id_code'")
            doms = comp.get("domains")
            if doms is None:
                errors.append(f"{prefix} missing 'domains' list")
            elif not isinstance(doms, list):
                errors.append(f"{prefix}.domains should be a list")
            else:
                for di, dom in enumerate(doms, start=1):
                    dprefix = f"{prefix}.domains[{di}]"
                    if not isinstance(dom, dict):
                        errors.append(f"{dprefix} is not a mapping")
                        continue
                    if "id" not in dom:
                        errors.append(f"{dprefix} missing 'id'")
                    if "id_code" not in dom:
                        errors.append(f"{dprefix} missing 'id_code'")
                    
                    if "mappings" in dom:
                        mappings = dom.get("mappings")
                        if not isinstance(mappings, dict):
                            errors.append(f"{dprefix}.mappings should be a mapping")
                        else:
                            for framework, refs in mappings.items():
                                if not isinstance(refs, list):
                                    errors.append(f"{dprefix}.mappings.{framework} should be a list")

                    if "maturity_levels" not in dom:
                        errors.append(f"{dprefix} missing 'maturity_levels' list")
                    else:
                        mls = dom.get("maturity_levels")
                        if not isinstance(mls, list):
                            errors.append(f"{dprefix}.maturity_levels should be a list")
                        else:
                            levels = [m.get("level") for m in mls if isinstance(m, dict)]
                            if sorted(levels) != [1,2,3,4,5]:
                                errors.append(f"{dprefix}.maturity_levels should contain levels 1..5")

    # Check questionnaire structure
    q = aismm.get("assessment_questionnaire")
    if q is None:
        errors.append("Missing 'assessment_questionnaire' section")
    else:
        questions = q.get("questions") if isinstance(q, dict) else None
        if questions is None:
            errors.append("'assessment_questionnaire' missing 'questions' list")
        elif not isinstance(questions, list):
            errors.append("assessment_questionnaire.questions should be a list")
        else:
            for qi, qq in enumerate(questions, start=1):
                qpre = f"assessment_questionnaire.questions[{qi}]"
                if not isinstance(qq, dict):
                    errors.append(f"{qpre} is not a mapping")
                    continue
                for field in ("id","type","title","help_text","required"):
                    if field not in qq:
                        errors.append(f"{qpre} missing '{field}'")
                allowed_types = ("multiple_choice","true_false","scoring","free_text","numeric")
                if qq.get("type") not in allowed_types:
                    errors.append(f"{qpre} has unsupported type '{qq.get('type')}'")

    return errors


def print_summary_v1(doc):
    """Print summary for v1.x structure."""
    pillars = doc.get("pillars", {})
    domains = doc.get("domains", {})
    total_questions = sum(len(d.get("questions", [])) for d in domains.values() if isinstance(d, dict))
    total_key_controls = sum(len(d.get("key_controls", [])) for d in domains.values() if isinstance(d, dict))
    
    print(f"Pillars: {len(pillars)}")
    print(f"Domains: {len(domains)}")
    print(f"Total questions (embedded): {total_questions}")
    print(f"Total key_controls: {total_key_controls}")
    
    # Check for framework mappings
    has_mitre = any("mitre_atlas" in str(d) for d in domains.values())
    has_owasp = any("owasp_genai" in str(d) for d in domains.values())
    print(f"MITRE ATLAS mappings: {'Yes' if has_mitre else 'No'}")
    print(f"OWASP GenAI mappings: {'Yes' if has_owasp else 'No'}")


def print_summary_v2(doc):
    """Print summary for v2.x structure."""
    aismm = doc.get("aismm", {})
    comps = aismm.get("components", [])
    total_domains = sum(len(c.get("domains", [])) for c in comps if isinstance(c, dict))
    q = aismm.get("assessment_questionnaire", {})
    qs = q.get("questions", [])
    
    print(f"Components: {len(comps)}")
    print(f"Total domains: {total_domains}")
    print(f"Total assessment questions: {len(qs)}")


def main():
    if not os.path.exists(YAML_PATH):
        print(f"ERROR: YAML file not found at {YAML_PATH}")
        sys.exit(2)
    
    try:
        doc = load_yaml(YAML_PATH)
    except Exception as e:
        print("ERROR: Failed to parse YAML:", e)
        sys.exit(2)
    
    version = detect_version(doc)
    if version is None:
        print("ERROR: Unable to detect AISMM version. Expected either 'aismm' key (v2.x) or 'domains'+'pillars' keys (v1.x)")
        sys.exit(2)
    
    print(f"Detected AISMM structure version: {version}")
    
    if version == "1.x":
        errors = validate_v1(doc)
    else:
        errors = validate_v2(doc)
    
    if errors:
        print(f"\nValidation FAILED with {len(errors)} issue(s):")
        for e in errors:
            print(" -", e)
        sys.exit(1)
    
    print("\n✓ YAML parsed and structure validated successfully.\n")
    
    if version == "1.x":
        print_summary_v1(doc)
    else:
        print_summary_v2(doc)


if __name__ == "__main__":
    main()

File: webapp/scripts/validate_yaml.ts
#!/usr/bin/env npx tsx
/**
 * YAML validation script for AISMM model definitions.
 * Supports both:
 *   - v1.x structure (flat domains dictionary, pillars, scoring_config)
 *   - v2.x structure (aismm wrapper with nested components)
 */

import { readFileSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import { parse as parseYaml } from 'yaml';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const ROOT = join(__dirname, '..', '..');
const YAML_PATH = join(ROOT, 'aismm_definition', 'aismm.yaml');

// Type definitions for YAML structures
interface V1Document {
  version?: string;
  name?: string;
  description?: string;
  scoring_config?: Record<string, unknown>;
  pillars?: Record<string, V1Pillar>;
  domains?: Record<string, V1Domain>;
}

interface V1Pillar {
  id?: string;
  name?: string;
  description?: string;
  weight?: number;
}

interface V1Domain {
  id?: string;
  name?: string;
  description?: string;
  pillar?: string;
  levels?: Record<string, unknown>;
  framework_alignment?: unknown[];
  key_controls?: unknown[];
  questions?: V1Question[];
}

interface V1Question {
  id?: string;
  text?: string;
  question_type?: string;
}

interface V2Document {
  aismm?: {
    components?: V2Component[];
    assessment_questionnaire?: {
      questions?: V2Question[];
    };
  };
}

interface V2Component {
  id?: string;
  id_code?: string;
  domains?: V2Domain[];
}

interface V2Domain {
  id?: string;
  id_code?: string;
  mappings?: Record<string, unknown[]>;
  maturity_levels?: Array<{ level?: number }>;
}

interface V2Question {
  id?: string;
  type?: string;
  title?: string;
  help_text?: string;
  required?: boolean;
}

/**
 * Load YAML file, rejecting files with tab characters.
 */
function loadYaml(path: string): unknown {
  const text = readFileSync(path, 'utf-8');
  
  if (text.includes('\t')) {
    const linesWithTabs = text.split('\n')
      .map((line, i) => line.includes('\t') ? i + 1 : null)
      .filter((n): n is number => n !== null);
    throw new Error(
      `YAML file contains tab characters on line(s): ${linesWithTabs.join(', ')}. ` +
      'YAML must use spaces for indentation. Replace tabs with spaces and re-run validation.'
    );
  }
  
  return parseYaml(text);
}

/**
 * Detect which AISMM structure version is being used.
 */
function detectVersion(doc: unknown): '1.x' | '2.x' | null {
  if (!doc || typeof doc !== 'object') {
    return null;
  }
  
  const record = doc as Record<string, unknown>;
  
  // v2.x has 'aismm' top-level key with 'components' inside
  if ('aismm' in record) {
    return '2.x';
  }
  
  // v1.x has 'domains', 'pillars', 'scoring_config' at top level
  if ('domains' in record && 'pillars' in record) {
    return '1.x';
  }
  
  return null;
}

/**
 * Validate v1.x structure (flat domains, pillars, scoring_config).
 */
function validateV1(doc: V1Document): string[] {
  const errors: string[] = [];
  
  // Check required top-level keys
  const requiredKeys = ['version', 'name', 'description', 'scoring_config', 'pillars', 'domains'];
  for (const key of requiredKeys) {
    if (!(key in doc)) {
      errors.push(`Missing required top-level key: '${key}'`);
    }
  }
  
  // Validate scoring_config
  const sc = doc.scoring_config;
  if (sc && typeof sc !== 'object') {
    errors.push("'scoring_config' should be a mapping");
  } else if (sc) {
    for (const key of ['level_scores', 'maturity_thresholds']) {
      if (!(key in sc)) {
        errors.push(`scoring_config missing '${key}'`);
      }
    }
  }
  
  // Validate pillars
  const pillars = doc.pillars || {};
  if (typeof pillars !== 'object') {
    errors.push("'pillars' should be a mapping");
  } else {
    for (const [pid, pdata] of Object.entries(pillars)) {
      if (!pdata || typeof pdata !== 'object') {
        errors.push(`pillars.${pid} should be a mapping`);
        continue;
      }
      for (const field of ['id', 'name', 'description', 'weight']) {
        if (!(field in pdata)) {
          errors.push(`pillars.${pid} missing '${field}'`);
        }
      }
    }
  }
  
  // Validate domains
  const domains = doc.domains || {};
  if (typeof domains !== 'object') {
    errors.push("'domains' should be a mapping");
  } else {
    for (const [did, ddata] of Object.entries(domains)) {
      const prefix = `domains.${did}`;
      if (!ddata || typeof ddata !== 'object') {
        errors.push(`${prefix} should be a mapping`);
        continue;
      }
      
      // Required domain fields
      for (const field of ['id', 'name', 'description', 'pillar']) {
        if (!(field in ddata)) {
          errors.push(`${prefix} missing '${field}'`);
        }
      }
      
      // Validate pillar reference
      const pillarRef = ddata.pillar;
      if (pillarRef && !(pillarRef in pillars)) {
        errors.push(`${prefix}.pillar '${pillarRef}' not found in pillars definition`);
      }
      
      // Validate levels
      const levels = ddata.levels || {};
      if (typeof levels !== 'object') {
        errors.push(`${prefix}.levels should be a mapping`);
      } else {
        const levelKeys = new Set(Object.keys(levels));
        const expectedLevels = new Set(['level_1', 'level_2', 'level_3', 'level_4', 'level_5']);
        
        const missing = [...expectedLevels].filter(l => !levelKeys.has(l));
        const extra = [...levelKeys].filter(l => !expectedLevels.has(l));
        
        if (missing.length > 0) {
          errors.push(`${prefix}.levels missing: ${missing.join(', ')}`);
        }
        if (extra.length > 0) {
          errors.push(`${prefix}.levels has unexpected keys: ${extra.join(', ')}`);
        }
      }
      
      // Validate framework_alignment (optional)
      const fa = ddata.framework_alignment;
      if (fa && !Array.isArray(fa)) {
        errors.push(`${prefix}.framework_alignment should be a list`);
      }
      
      // Validate key_controls (optional)
      const kc = ddata.key_controls;
      if (kc && !Array.isArray(kc)) {
        errors.push(`${prefix}.key_controls should be a list`);
      }
      
      // Validate questions
      const questions = ddata.questions;
      if (questions && !Array.isArray(questions)) {
        errors.push(`${prefix}.questions should be a list`);
      } else if (questions) {
        questions.forEach((q, qi) => {
          const qprefix = `${prefix}.questions[${qi + 1}]`;
          if (!q || typeof q !== 'object') {
            errors.push(`${qprefix} should be a mapping`);
            return;
          }
          for (const field of ['id', 'text', 'question_type']) {
            if (!(field in q)) {
              errors.push(`${qprefix} missing '${field}'`);
            }
          }
        });
      }
    }
  }
  
  return errors;
}

/**
 * Validate v2.x structure (aismm wrapper with nested components).
 */
function validateV2(doc: V2Document): string[] {
  const errors: string[] = [];
  
  if (!('aismm' in doc)) {
    errors.push("Missing top-level 'aismm' key");
    return errors;
  }
  
  const aismm = doc.aismm;
  if (!aismm || typeof aismm !== 'object') {
    errors.push("'aismm' should be a mapping");
    return errors;
  }
  
  const comps = aismm.components;
  if (comps === undefined) {
    errors.push("Missing 'components' list");
  } else if (!Array.isArray(comps)) {
    errors.push("'components' should be a list");
  } else {
    comps.forEach((comp, ci) => {
      const prefix = `components[${ci + 1}]`;
      if (!comp || typeof comp !== 'object') {
        errors.push(`${prefix} is not a mapping`);
        return;
      }
      if (!('id' in comp)) {
        errors.push(`${prefix} missing 'id'`);
      }
      if (!('id_code' in comp)) {
        errors.push(`${prefix} missing 'id_code'`);
      }
      
      const doms = comp.domains;
      if (doms === undefined) {
        errors.push(`${prefix} missing 'domains' list`);
      } else if (!Array.isArray(doms)) {
        errors.push(`${prefix}.domains should be a list`);
      } else {
        doms.forEach((dom, di) => {
          const dprefix = `${prefix}.domains[${di + 1}]`;
          if (!dom || typeof dom !== 'object') {
            errors.push(`${dprefix} is not a mapping`);
            return;
          }
          if (!('id' in dom)) {
            errors.push(`${dprefix} missing 'id'`);
          }
          if (!('id_code' in dom)) {
            errors.push(`${dprefix} missing 'id_code'`);
          }
          
          if ('mappings' in dom) {
            const mappings = dom.mappings;
            if (!mappings || typeof mappings !== 'object') {
              errors.push(`${dprefix}.mappings should be a mapping`);
            } else {
              for (const [framework, refs] of Object.entries(mappings)) {
                if (!Array.isArray(refs)) {
                  errors.push(`${dprefix}.mappings.${framework} should be a list`);
                }
              }
            }
          }
          
          if (!('maturity_levels' in dom)) {
            errors.push(`${dprefix} missing 'maturity_levels' list`);
          } else {
            const mls = dom.maturity_levels;
            if (!Array.isArray(mls)) {
              errors.push(`${dprefix}.maturity_levels should be a list`);
            } else {
              const levels = mls
                .filter((m): m is { level: number } => typeof m === 'object' && m !== null && 'level' in m)
                .map(m => m.level)
                .sort((a, b) => a - b);
              
              if (JSON.stringify(levels) !== JSON.stringify([1, 2, 3, 4, 5])) {
                errors.push(`${dprefix}.maturity_levels should contain levels 1..5`);
              }
            }
          }
        });
      }
    });
  }
  
  // Check questionnaire structure
  const q = aismm.assessment_questionnaire;
  if (q === undefined) {
    errors.push("Missing 'assessment_questionnaire' section");
  } else {
    const questions = typeof q === 'object' && q !== null ? q.questions : undefined;
    if (questions === undefined) {
      errors.push("'assessment_questionnaire' missing 'questions' list");
    } else if (!Array.isArray(questions)) {
      errors.push("assessment_questionnaire.questions should be a list");
    } else {
      const allowedTypes = ['multiple_choice', 'true_false', 'scoring', 'free_text', 'numeric'];
      
      questions.forEach((qq, qi) => {
        const qpre = `assessment_questionnaire.questions[${qi + 1}]`;
        if (!qq || typeof qq !== 'object') {
          errors.push(`${qpre} is not a mapping`);
          return;
        }
        for (const field of ['id', 'type', 'title', 'help_text', 'required']) {
          if (!(field in qq)) {
            errors.push(`${qpre} missing '${field}'`);
          }
        }
        if (qq.type && !allowedTypes.includes(qq.type)) {
          errors.push(`${qpre} has unsupported type '${qq.type}'`);
        }
      });
    }
  }
  
  return errors;
}

/**
 * Print summary for v1.x structure.
 */
function printSummaryV1(doc: V1Document): void {
  const pillars = doc.pillars || {};
  const domains = doc.domains || {};
  
  const totalQuestions = Object.values(domains)
    .filter((d): d is V1Domain => typeof d === 'object' && d !== null)
    .reduce((sum, d) => sum + (d.questions?.length || 0), 0);
  
  const totalKeyControls = Object.values(domains)
    .filter((d): d is V1Domain => typeof d === 'object' && d !== null)
    .reduce((sum, d) => sum + (d.key_controls?.length || 0), 0);
  
  console.log(`Pillars: ${Object.keys(pillars).length}`);
  console.log(`Domains: ${Object.keys(domains).length}`);
  console.log(`Total questions (embedded): ${totalQuestions}`);
  console.log(`Total key_controls: ${totalKeyControls}`);
  
  // Check for framework mappings
  const domainStr = JSON.stringify(domains);
  const hasMitre = domainStr.includes('mitre_atlas');
  const hasOwasp = domainStr.includes('owasp_genai');
  console.log(`MITRE ATLAS mappings: ${hasMitre ? 'Yes' : 'No'}`);
  console.log(`OWASP GenAI mappings: ${hasOwasp ? 'Yes' : 'No'}`);
}

/**
 * Print summary for v2.x structure.
 */
function printSummaryV2(doc: V2Document): void {
  const aismm = doc.aismm || {};
  const comps = aismm.components || [];
  
  const totalDomains = comps
    .filter((c): c is V2Component => typeof c === 'object' && c !== null)
    .reduce((sum, c) => sum + (c.domains?.length || 0), 0);
  
  const q = aismm.assessment_questionnaire || {};
  const qs = q.questions || [];
  
  console.log(`Components: ${comps.length}`);
  console.log(`Total domains: ${totalDomains}`);
  console.log(`Total assessment questions: ${qs.length}`);
}

/**
 * Main function
 */
function main(): void {
  let doc: unknown;
  
  try {
    doc = loadYaml(YAML_PATH);
  } catch (e) {
    console.error('ERROR: Failed to parse YAML:', (e as Error).message);
    process.exit(2);
  }
  
  const version = detectVersion(doc);
  if (version === null) {
    console.error("ERROR: Unable to detect AISMM version. Expected either 'aismm' key (v2.x) or 'domains'+'pillars' keys (v1.x)");
    process.exit(2);
  }
  
  console.log(`Detected AISMM structure version: ${version}`);
  
  let errors: string[];
  if (version === '1.x') {
    errors = validateV1(doc as V1Document);
  } else {
    errors = validateV2(doc as V2Document);
  }
  
  if (errors.length > 0) {
    console.log(`\nValidation FAILED with ${errors.length} issue(s):`);
    for (const e of errors) {
      console.log(' -', e);
    }
    process.exit(1);
  }
  
  console.log('\n✓ YAML parsed and structure validated successfully.\n');
  
  if (version === '1.x') {
    printSummaryV1(doc as V1Document);
  } else {
    printSummaryV2(doc as V2Document);
  }
}

main();
